### PR TITLE
Change skip_unavailable default value to true

### PR DESCRIPTION
**Pull request by quux00**
In order to improve the experience of cross-cluster search, we propose changing
the default value of the remote cluster `skip_unavailable` setting from `false` to `true`.

This setting causes any cross-cluster search to entirely fail when any remote cluster
with `skip_unavailable=false` is either unavailable (connection to it fails) or if the
search on it fails on all shards.

Setting `skip_unavailable=true` allows partial results from other clusters to be
returned. In that case, the search response cluster metadata will show a `skipped`
status, so the user can see that no data came in from that cluster. Kibana also
now leverages this metadata in the cross-cluster search responses to allow users
to see how many clusters returned data and drill down into which clusters did not
(including failure messages).

Currently, the user/admin has to specifically set the value to `true` in the configs, like so:

```
cluster:
    remote:
        remote1:
            seeds: 10.10.10.10:9300
            skip_unavailable: true
```

even though that is probably what search admins want in the vast majority of cases.

Setting `skip_unavailable=false` should be a conscious (and probably rare) choice
by an Elasticsearch admin that a particular cluster's results are so essential to a
search (or visualization in dashboard or Discover panel) that no results at all should
be shown if it cannot return any results.

### CODE REVIEW

**Comment by quux00** [comment]
I have a PR out for this change now: https://github.com/elastic/elasticsearch/pull/105792

In the process of doing this, I found some potential issues, both of which are pre-existing but now perhaps become more acute, so they may need to be addressed.

**(1)** For `_search` (and `_async_search`), when `skip_unavailable`=`false` and a cluster has an issue (such as a Security/permissions exception or IndexNotFound), that error is explicitly returned, along with an error status code such as 404 or 403. As expected, when `skip_unavailable`=`true` and the same issue occurs, an HTTP status of 200 is returned, but the error reported in the SearchResponse is `NoSuchRemoteClusterException`, not the  `IndexNotFoundException`. This feels like a bug in the cluster resolution logic (es-distributed side?).

**(2)** The above also occurs for an internal API [RemoteClusterNodesAction], an API I wasn't aware of. I have a question out to #es-security about whether this is a concern and whether an issue should be filed: https://elastic.slack.com/archives/CA6N3HDF1/p1709675350567369

**(3)** The `reindex` API is also affected by the `skip_unavailable` when doing a remote reindexing operation. When  `skip_unavailable`=`false` and a reindex targets an index that does not exist, then `IndexNotFoundException` is returned with an HTTP status error code.

However, when `skip_unavailable`=`true` it does not return an error status code, and worse it does not report any error in the response object:

```json
{
    "took": 36,
    "timed_out": false,
    "total": 0,
    "updated": 0,
    "created": 0,
    "deleted": 0,
    "batches": 0,
    "version_conflicts": 0,
    "noops": 0,
    "retries": {
        "bulk": 0,
        "search": 0
    },
    "throttled_millis": 0,
    "requests_per_second": -1,
    "throttled_until_millis": 0,
    "failures": []
}
```

Should the reindex API report an issue in the `failures` section of this response in this case? (It likely needs to parse the search response against the remote cluster.) Or should it return an error status code? Should the reindex API have a dependency on the `skip_unavailable` setting?

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java [hunk 1]
```diff
@@ -87,7 +87,7 @@ public final class RemoteClusterService extends RemoteClusterAware implements Cl
     public static final Setting.AffixSetting<Boolean> REMOTE_CLUSTER_SKIP_UNAVAILABLE = Setting.affixKeySetting(
         "cluster.remote.",
         "skip_unavailable",
-        (ns, key) -> boolSetting(key, false, new RemoteConnectionEnabled<>(ns, key), Setting.Property.Dynamic, Setting.Property.NodeScope)
+        (ns, key) -> boolSetting(key, true, new RemoteConnectionEnabled<>(ns, key), Setting.Property.Dynamic, Setting.Property.NodeScope)
```
Comment (quux00) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java:90]: This is the only production code change. Everything else is tests and docs.

**Comment by quux00** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Review by mark-vieira — APPROVED** [review]

File: qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml [hunk 2]
```diff
@@ -152,7 +152,7 @@
   - do:
       cluster.remote_info: {}
 
-  - is_false: remote1.skip_unavailable
+  - is_true: remote1.skip_unavailable
```
Comment (javanna) [inline qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml:155]: I guess these yaml tests are run only against the latest versions and never cross version?

File: docs/reference/modules/cluster/remote-clusters-settings.asciidoc [hunk 3]
```diff
@@ -28,9 +28,14 @@ mode are described separately.
...
-  cluster request. Default is `false`, meaning that all clusters are mandatory
-  by default, but they can selectively be made optional by setting this setting
-  to `true`.
+  cluster request.
+
+IMPORTANT: In Elasticsearch 8.15, the default value for `skip_unavailable` was
+changed from `false` to `true`. Before Elasticsearch 8.15, if you want a cluster
+to be treated as optional for a {ccs}, then you need to set that configuration.
+From Elasticsearch 8.15 forward, you need to set the configuration in order to
+make a cluster required for the {ccs}.
```
Comment (javanna) [inline docs/reference/modules/cluster/remote-clusters-settings.asciidoc:37]: Maybe obvious, but should we be explicit about the upgrade path: once you upgrade the coordinating node / cluster, where the remotes are registered, you get the new behavior? Say that the coord cluster has multiple nodes that are upgraded one by one (rolling upgrade) what would be the behavior there?

**Review by javanna — APPROVED** [review]
I left a question about yaml tests and the upgrade path, LGTM otherwise

File: qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml (hunk 2 above)
Comment (quux00) [reply qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml:155]: I think so? The build hasn't been failing on bwc tests. If it does at some point, we can just remove this first test block that tests the default and require the value to be explicitly set in the yaml tests.

File: docs/reference/modules/cluster/remote-clusters-settings.asciidoc (hunk 3 above)
Comment (quux00) [reply docs/reference/modules/cluster/remote-clusters-settings.asciidoc:37]: Good point, but that is very specific to a one-time upgrade, so should that go in the "permanent" API docs? Or should it go into some one-time doc like the changelog or deprecation announcement?

File: qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml (hunk 2 above)
Comment (javanna) [reply qa/multi-cluster-search/src/test/resources/rest-api-spec/test/multi_cluster/20_info.yml:155]: This does make me wonder if we are lacking tests. Do we have any that verify the behavior with clusters from different versions?

File: docs/reference/modules/cluster/remote-clusters-settings.asciidoc (hunk 3 above)
Comment (quux00) [reply docs/reference/modules/cluster/remote-clusters-settings.asciidoc]: I added a blurb here about that and we'll also include in the 8.15 release notes. Thanks for flagging this.

**Comment by quux00** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**